statsmodels/regression/tests/test_lme.py

```diff
@@ -1057,99 +1104,125 @@ def test_random_effects_getters():
-    assert(np.corrcoef(b0[:, 1], b[:, 1])[0, 1] > 0.8)
+    assert (np.corrcoef(b0[:, 0], b[:, 0])[0, 1] > 0.8)
+    assert (np.corrcoef(b0[:, 1], b[:, 1])[0, 1] > 0.8)
 
     cf0 = [ref[k][2:6] for k in range(ng)]
     cf0 = np.asarray(cf0)
     for k in range(4):
-        assert(np.corrcoef(cf0[:, k], cc[:, k])[0, 1] > 0.8)
+        assert (np.corrcoef(cf0[:, k], cc[:, k])[0, 1] > 0.8)
 
     # Smoke test for predictive covariances
     refc = result.random_effects_cov
     for g in refc.keys():
         p = ref[g].size
-        assert(refc[g].shape == (p, p))
+        assert (refc[g].shape == (p, p))
 
 
 def test_smw_solver():
 
     np.random.seed(23)
-    p = 5
-    q = 4
-    r = 2
-    s = 2
 
-    A = np.random.normal(size=(p, q))
-    AtA = np.dot(A.T, A)
+    def tester(p, q, r, s):
+
+        d = q - r
+
+        A = np.random.normal(size=(p, q))
+        AtA = np.dot(A.T, A)
+
+        B = np.zeros((q, q))
+        B[0:r, 0:r] = np.random.normal(size=(r, r))
+        di = np.random.uniform(size=d)
+        B[r:q, r:q] = np.diag(1 / di)
+        Qi = np.linalg.inv(B[0:r, 0:r])
+        s = 0.5
 
-    B = np.zeros((q, q))
-    B[0:r, 0:r] = np.random.normal(size=(r, r))
-    di = np.random.uniform(size=s)
-    B[r:q, r:q] = np.diag(1/di)
-    Qi = np.linalg.inv(B[0:r, 0:r])
-    s = 0.5
+        x = np.random.normal(size=p)
+        y2 = np.linalg.solve(s * np.eye(p, p) + np.dot(A, np.dot(B, A.T)), x)
 
-    x = np.random.normal(size=p)
-    y2 = np.linalg.solve(s*np.eye(p, p) + np.dot(A, np.dot(B, A.T)), x)
+        f = _smw_solver(s, A, AtA, Qi, di)
+        y1 = f(x)
+        assert_allclose(y1, y2)
 
-    f = _smw_solver(s, A, AtA, Qi, di)
-    y1 = f(x)
-    assert_allclose(y1, y2)
+        f = _smw_solver(s, sparse.csr_matrix(A), sparse.csr_matrix(AtA), Qi,
+                        di)
+        y1 = f(x)
+        assert_allclose(y1, y2)
 
-    f = _smw_solver(s, sparse.csr_matrix(A), sparse.csr_matrix(AtA), Qi, di)
-    y1 = f(x)
-    assert_allclose(y1, y2)
+    for p in 5, 10:
+        for q in 4, 8:
+            for r in 2, 3:
+                for s in 0, 0.5:
+                    tester(p, q, r, s)
 
 
 def test_smw_logdet():
 
     np.random.seed(23)
-    p = 5
-    q = 4
-    r = 2
-    s = 2
-
-    A = np.random.normal(size=(p, q))
-    AtA = np.dot(A.T, A)
-
-    B = np.zeros((q, q))
-    c = np.random.normal(size=(r, r))
-    B[0:r, 0:r] = np.dot(c.T, c)
-    di = np.random.uniform(size=s)
-    B[r:q, r:q] = np.diag(1/di)
-    Qi = np.linalg.inv(B[0:r, 0:r])
-    s = 0.5
-
-    _, d2 = np.linalg.slogdet(s*np.eye(p, p) + np.dot(A, np.dot(B, A.T)))
-
-    _, bd = np.linalg.slogdet(B)
-    d1 = _smw_logdet(s, A, AtA, Qi, di, bd)
-    assert_allclose(d1, d2)
+
+    def tester(p, q, r, s):
+
+        d = q - r
+        A = np.random.normal(size=(p, q))
+        AtA = np.dot(A.T, A)
+
+        B = np.zeros((q, q))
+        c = np.random.normal(size=(r, r))
+        B[0:r, 0:r] = np.dot(c.T, c)
```

Comment:
The other `tester` func sets `B[0:r, 0:r] = c` (effectively).  Is the difference intentional, and if not, can these share some code?

Reply:
The construction of B needs to be different.  In the log determinant calculation, B needs to be positive definite, but for the solver B only needs to be non-singular.